src/Microsoft.AspNet.OData.Shared/Formatter/Serialization/ODataResourceSerializer.cs

```diff
@@ -634,6 +638,109 @@ public virtual ODataResource CreateResource(SelectExpandNode selectExpandNode, R
+                {
+                    if (resource.InstanceAnnotations == null)
+                    {
+                        resource.InstanceAnnotations = new List<ODataInstanceAnnotation>();
+                    }
+
+                    foreach (KeyValuePair<string,object> annotation in instanceAnnotationsDictionary[string.Empty])
+                    {
+                        IEdmTypeReference edmTypeReference = resourceContext.SerializerContext.GetEdmType(annotation.Value,
+                           annotation.Value.GetType());
+                         
+                        ODataEdmTypeSerializer edmTypeSerializer = GetEdmTypeSerializer(edmTypeReference);
+
+                        if (edmTypeSerializer != null)
+                        {
+                            ODataValue annotationValue = edmTypeSerializer.CreateODataValue(annotation.Value, edmTypeReference, resourceContext.SerializerContext);
+
+                            if (annotationValue != null)
+                            {
+                                resource.InstanceAnnotations.Add(new ODataInstanceAnnotation(annotation.Key, annotationValue));
+                            }
+                        }
+                    }
+                }
+                
+                foreach(ODataProperty property in resource.Properties)
+                {
+                    string propertyName = property.Name;
+
+                    if (instanceAnnotationsDictionary.ContainsKey(propertyName))
+                    {
+                        if (property.InstanceAnnotations == null)
+                        {
+                            property.InstanceAnnotations = new List<ODataInstanceAnnotation>();
+                        }
+
+                        foreach (KeyValuePair<string,object> annotation in instanceAnnotationsDictionary[propertyName])
+                        {
+                            IEdmTypeReference edmTypeReference = resourceContext.SerializerContext.GetEdmType(annotation.Value,
+                            annotation.Value.GetType());
+
+                            ODataEdmTypeSerializer edmTypeSerializer = GetEdmTypeSerializer(edmTypeReference);
+                            if (edmTypeSerializer != null)
+                            {
+                                ODataValue annotationValue = edmTypeSerializer.CreateODataValue(annotation.Value, edmTypeReference, resourceContext.SerializerContext);
+
+                                if (annotationValue != null)
+                                {
+                                    property.InstanceAnnotations.Add(new ODataInstanceAnnotation(annotation.Key, annotationValue));
+                                }
+                            }
+                        }
+                    }
+                } 
+            }
+
+        }
+
+        private ODataEdmTypeSerializer GetEdmTypeSerializer(IEdmTypeReference edmTypeReference)
```

Comment:
>GetEdmTypeSerializer [](start = 39, length = 20)

you should retrieve the serializer from the DI in "ODataSerializerProvider"?

Reply:
Here the common GetEdmTypeSerializer in ODataSerializerProvider has a different implementation, especially for complex types